File: Sample/BbQ.ChatWidgets.Sample.Shared/EChartsWidget.cs
using System.Text.Json.Serialization;
using BbQ.ChatWidgets.Models;

namespace BbQ.ChatWidgets.Sample.Shared;

/// <summary>
/// A custom ECharts widget demonstrating extensibility without modifying the core library.
/// Renders interactive charts using the Apache ECharts library.
/// </summary>
public sealed record EChartsWidget(
    string Label,
    string Action,
    string ChartType,      // 'bar', 'line', 'pie', etc.
    string JsonData        // Raw ECharts options JSON
) : ChatWidget(Label, Action)
{
    [JsonIgnore]
    public override string Purpose =>
        """
        ***EChart Widget***
        Emitted through the schema-constrained emit_widgets tool as complete chart state.
        Renders interactive charts using Apache ECharts.
        Use when you need to visualize data interactively within the chat interface.
        Note: The 'jsonData' field must contain valid ECharts options in JSON string.
        The 'chartType' field specifies the type of chart to render (e.g., 'bar', 'line', 'pie').
        """;
}

File: Sample/BbQ.ChatWidgets.Sample.Angular/ClientApp/src/app/widgets/EChartsWidget.ts
import { ChatWidget } from '@bbq-chat/widgets';

/**
 * Custom ECharts widget demonstrating extensibility.
 * Renders interactive charts using Apache ECharts library.
 * 
 * Example usage:
 * new EChartsWidget(
 *   "Sales Chart",
 *   "on_chart_click",
 *   "bar",
 *   JSON.stringify({...echartsOptions...})
 * )
 */
export class EChartsWidget extends ChatWidget {
  public readonly chartType: string;
  public readonly jsonData: string;

  constructor(
    label: string,
    action: string,
    chartType: string,
    jsonData: string
  ) {
    super('echarts', label, action);
    this.chartType = chartType;
    this.jsonData = jsonData;
  }

  toObject(): any {
    const base = this as any;
    return {
      type: 'echarts',
      label: base.label,
      action: base.action,
      chartType: this.chartType,
      jsonData: this.jsonData,
    };
  }
}
